linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -1538,25 +1538,30 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
 
       def doReturnToLabel(l: LabelIdent): js.Tree = {
         val newLhs = env.lhsForLabeledExpr(l)
-        val body = pushLhsInto(newLhs, rhs, Set.empty)
         if (newLhs.hasNothingType) {
           /* A touch of peephole dead code elimination.
            * This is actually necessary to avoid dangling breaks to eliminated
            * labels, as in issue #2307.
            */
-          body
+          pushLhsInto(newLhs, rhs, tailPosLabels)
         } else if (tailPosLabels.contains(l.name)) {
-          body
-        } else if (env.isDefaultBreakTarget(l.name)) {
-          js.Block(body, js.Break(None))
-        } else if (env.isDefaultContinueTarget(l.name)) {
-          js.Block(body, js.Continue(None))
+          pushLhsInto(newLhs, rhs, tailPosLabels)
         } else {
-          usedLabels += l.name
-          val transformedLabel = Some(transformLabelIdent(l))
-          val jump =
-            if (env.isLabelTurnedIntoContinue(l.name)) js.Continue(transformedLabel)
-            else js.Break(transformedLabel)
+          val body = pushLhsInto(newLhs, rhs, Set.empty)
+
+          val jump = if (env.isDefaultBreakTarget(l.name)) {
+            js.Break(None)
+          } else if (env.isDefaultContinueTarget(l.name)) {
+            js.Continue(None)
+          } else {
+            usedLabels += l.name
+            val transformedLabel = Some(transformLabelIdent(l))
+            if (env.isLabelTurnedIntoContinue(l.name))
+              js.Continue(transformedLabel)
+            else
+              js.Break(transformedLabel)
+          }
+
```

Comment:
I don't understand how these changes are related to the return change. IIUC pushing down the `tailPosLabels` could/should be done either way?

Reply:
These changes were necessary to maintain the status quo in terms of the emitted JS. My understanding is that before, we would get a lot of `Block(stats, Return(Undefined))` which are now `Return(Block(stats))`. If the `stats` contain `If` and other branches, the new `Return` can be pushed down a lot more. When it wasn't pushed down, we would have regular fall-through behavior without `js.Break`s. Now with the `Return`s pushed down, we would get a lot of `js.Break`s that were not actually necessary because of the missing optimization being done here.

Reply:
OK, that makes sense. Maybe we can add this explanation to the commit message so we have a record of it?